### PR TITLE
Dev i2c #379

### LINKED ISSUE #379: Opening I2C device: Too many open files
we are opening I2C device files without closing them. 

```
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:  2023-08-08 10:37:52,188 INFO [device_shadow.py:449] Applying shadow state to device DAC
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:  2023-08-08 10:37:52,188 INFO [edgepi_dac.py:57] Initializing DAC Bus
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:  2023-08-08 10:37:52,189 INFO [edgepi_eeprom.py:40] Initializing EEPROM Access
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:  2023-08-08 10:37:52,189 ERROR [device_shadow.py:491] Could not apply state to DAC
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:  2023-08-08 10:37:52,189 ERROR [device_shadow.py:492] [Errno 24] Opening I2C device: Too many open files
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]: Traceback (most recent call last):
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:   File "/opt/edgepi/edgepi-device-manager/periphery/i2c.py", line 72, in _open
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]: OSError: [Errno 24] Too many open files: '/dev/i2c-10'
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]: During handling of the above exception, another exception occurred:
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]: Traceback (most recent call last):
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:   File "/opt/edgepi/edgepi-device-manager/device_shadow/device_shadow.py", line 467, in __apply_shadow_state
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:   File "/opt/edgepi/edgepi-device-manager/device_shadow/device_shadow.py", line 243, in __get_update_handler
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:   File "/opt/edgepi/edgepi-device-manager/module/update_handlers/update_handler_factory.py", line 111, in get
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:   File "/opt/edgepi/edgepi-device-manager/module/update_handlers/update_handler_factory.py", line 298, in __get_update_handler_outputs_module
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:   File "/opt/edgepi/edgepi-device-manager/module/update_handlers/update_handler_dac.py", line 79, in __init__
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:   File "/opt/edgepi/edgepi-device-manager/edgepi/dac/edgepi_dac.py", line 61, in __init__
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:   File "/opt/edgepi/edgepi-device-manager/edgepi/eeprom/edgepi_eeprom.py", line 45, in __init__
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:   File "/opt/edgepi/edgepi-device-manager/edgepi/peripherals/i2c.py", line 21, in __init__
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:   File "/opt/edgepi/edgepi-device-manager/periphery/i2c.py", line 58, in __init__
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:   File "/opt/edgepi/edgepi-device-manager/periphery/i2c.py", line 74, in _open
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]: periphery.i2c.I2CError: [Errno 24] Opening I2C device: Too many open files

```

similar issue on spi:
```
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:  2023-08-08 10:37:52,234 INFO [device_shadow.py:449] Applying shadow state to device DOUT
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:  2023-08-08 10:37:52,235 INFO [edgepi_gpio.py:26] GPIO initializing
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:  2023-08-08 10:37:52,236 INFO [edgepi_dac.py:57] Initializing DAC Bus
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:  2023-08-08 10:37:52,236 ERROR [device_shadow.py:491] Could not apply state to DOUT
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:  2023-08-08 10:37:52,237 ERROR [device_shadow.py:492] [Errno 24] Opening SPI device: Too many open files
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]: Traceback (most recent call last):
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:   File "/opt/edgepi/edgepi-device-manager/periphery/spi.py", line 109, in _open
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]: OSError: [Errno 24] Too many open files: '/dev/spidev6.3'
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]: During handling of the above exception, another exception occurred:
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]: Traceback (most recent call last):
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:   File "/opt/edgepi/edgepi-device-manager/device_shadow/device_shadow.py", line 467, in __apply_shadow_state
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:   File "/opt/edgepi/edgepi-device-manager/device_shadow/device_shadow.py", line 243, in __get_update_handler
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:   File "/opt/edgepi/edgepi-device-manager/module/update_handlers/update_handler_factory.py", line 111, in get
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:   File "/opt/edgepi/edgepi-device-manager/module/update_handlers/update_handler_factory.py", line 298, in __get_update_handler_outputs_module
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:   File "/opt/edgepi/edgepi-device-manager/module/update_handlers/update_handler_dout.py", line 79, in __init__
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:   File "/opt/edgepi/edgepi-device-manager/edgepi/digital_output/edgepi_digital_output.py", line 41, in __init__
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:   File "/opt/edgepi/edgepi-device-manager/edgepi/dac/edgepi_dac.py", line 58, in __init__
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:   File "/opt/edgepi/edgepi-device-manager/edgepi/peripherals/spi.py", line 34, in __init__
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:   File "/opt/edgepi/edgepi-device-manager/periphery/spi.py", line 75, in __init__
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]:   File "/opt/edgepi/edgepi-device-manager/periphery/spi.py", line 111, in _open
Aug 08 10:37:52 edgepi-r4p87y edgepi_device_manager[609]: periphery.spi.SPIError: [Errno 24] Opening SPI device: Too many open files
```

pwm issue might be because of the same problem.

### PR DESCRIPTION
**Pull request by sjpark608**
Closes #379

I2C/SPI refactored to use context manger to handle closing the device

### CODE REVIEW

File: src/edgepi/peripherals/spi.py [hunk 1]
```diff
@@ -31,20 +30,33 @@ def __init__(
     ):
         self.devpath = f"/dev/spidev{bus_num}.{dev_id}"
         _logger.debug(f"Initialized SPI device with path '{self.devpath}'")
```
Comment (farzadpanahi) [inline src/edgepi/peripherals/spi.py:32]: move spi init log to spi_open

**Review by farzadpanahi — DISMISSED** [review]

**Comment by farzadpanahi** [comment]
@sjpark608 as discussed, please add integration tests to blast i2c and spi devices over the limit of file descriptors to see if any issue happens.

File: src/edgepi/peripherals/spi.py (hunk 1 above)
Comment (sjpark608) [reply src/edgepi/peripherals/spi.py]: [b34b401](https://github.com/EdgePi-Cloud/edgepi-python-sdk/pull/380/commits/b34b401442f857301b121e460db547b050157600)

File: src/edgepi/peripherals/i2c.py [hunk 2]
```diff
@@ -16,9 +17,25 @@ class I2CDevice():
     I2C Device class
     '''
     def __init__(self, fd: str = None):
-        self.fd = fd
-        _logger.debug(f"Initialized I2C device with path '{self.fd}'")
-        self.i2cdev = I2C(devpath=fd)
+        self.i2c_fd = fd
+        self.i2cdev = None
+        _logger.debug(f"Initialized I2C device with path '{self.i2c_fd}'")
```
[the remark this answers is not in the document]
Comment (sjpark608) [reply src/edgepi/peripherals/i2c.py]: [b34b401](https://github.com/EdgePi-Cloud/edgepi-python-sdk/pull/380/commits/b34b401442f857301b121e460db547b050157600)

**Review by farzadpanahi — APPROVED** [review]

**Pull request merged**